core/lib/multivm/src/dump.rs

```diff
@@ -0,0 +1,379 @@
+use std::{collections::HashMap, fmt, sync::Arc};
+
+use serde::{Deserialize, Serialize};
+use zksync_types::{
+    block::L2BlockExecutionData, web3, L1BatchNumber, L2BlockNumber, Transaction, H256, U256,
+};
+use zksync_utils::u256_to_h256;
+use zksync_vm_interface::storage::WriteStorage;
+
+use crate::{
+    interface::{
+        storage::{InMemoryStorage, ReadStorage, StoragePtr, StorageView},
+        BootloaderMemory, BytecodeCompressionError, CompressedBytecodeInfo, CurrentExecutionState,
+        FinishedL1Batch, L1BatchEnv, L2BlockEnv, SystemEnv, VmExecutionMode,
+        VmExecutionResultAndLogs, VmFactory, VmInterface, VmInterfaceHistoryEnabled,
+        VmMemoryMetrics,
+    },
+    vm_fast, vm_latest, HistoryMode,
+};
+
+/// VM that tracks decommitment of bytecodes during execution. This is required to create a [`VmDump`].
+pub trait VmTrackingContracts: VmInterface {
```

Comment:
It could make sense to make this a part of `VmInterface` trait (maybe, even replace `get_current_execution_state`, which has not quite obvious caveats, like being callable only at the end of VM execution in old VM versions). If so, can do this in a follow-up PR.